### ISSUE TITLE
Add Occuseg

### ISSUE THREAD

**Issue by nicolas-chaulet** [nicolas-chaulet/torch-points3d#249]
https://arxiv.org/pdf/2003.06537v3.pdf

**Comment by zetyquickly** [comment]
@nicolas-chaulet is it possible to decompose this on tasks? I will contribute

**Comment by nicolas-chaulet** [comment]
Hi @zetyquickly that sounds great! I think a good place to start would be to add the occupancy prediction and the spatial term (equations 10 and 3 in the paper). Maybe the spatial term is the easiest to start with, you could add it to the Minkowski model for semantic segmentation where each point predicts the class it belongs to but also the centre of the object it belongs to. You will have to do a few things:
- modify the semantic dataset (you can use Scannet) to return object instance labels and object centres 
- change the output layer to class and object centre
- implement the loss

Once this is done we can add the occupancy term, for that you can use the instance labels from the previous step as ground truth and every point needs to predict this number (see paper for details)

Let me also add you to our Slack, it will be easier if you want to ask questions.

**Comment by lzhnb** [comment]
@nicolas-chaulet Would you like to share how you plan to implement the super-voxel clustering algorithm? Using pcl or others library, I am actually most interested in this for this work. Thanks!

**Comment by nicolas-chaulet** [comment]
I was more thinking of using scikit learn for that or some basic clustering toolbox. Adding PCL as a dependency is not a good idea in general and it comes with a large maintenance cost...

**Comment by lzhnb** [comment]
> I was more thinking of using scikit learn for that or some basic clustering toolbox. Adding PCL as a dependency is not a good idea in general and it comes with a large maintenance cost...

Oh, `pcl` is really verbose for your library, and the `segmentation.slic` in `skimage` maybe a good choice, you can try it, but compare with pcl, it is slow, and I think for occuseg, it's sensitive to the clustering parameters.

Looking forward for your implementation. Best wish!

**Comment by zetyquickly** [comment]
Can't we just use segmentation tool from ScanNet repository for this?

**Comment by lzhnb** [comment]
> Can't we just use segmentation tool from ScanNet repository for this?

Do you mean the [segmentator](https://github.com/ScanNet/ScanNet/tree/master/Segmentator)? It sounds good. And can it be used for other datasets？

**Comment by zetyquickly** [comment]
> > Can't we just use segmentation tool from ScanNet repository for this?
> 
> Do you mean the [segmentator](https://github.com/ScanNet/ScanNet/tree/master/Segmentator)? It sounds good. And can it be used for other datasets？I see its description, it can only work on mesh, point cloud maybe not.

Yes, I mean this very segmentator. I believe that original paper's method used that, I've seen the code once

**Comment by lzhnb** [comment]
> > > Can't we just use segmentation tool from ScanNet repository for this?
> > 
> > 
> > Do you mean the [segmentator](https://github.com/ScanNet/ScanNet/tree/master/Segmentator)? It sounds good. And can it be used for other datasets？I see its description, it can only work on mesh, point cloud maybe not.
> 
> Yes, I mean this very segmentator. I believe that original paper's method used that, I've seen the code once

Really? I think the paper means the super-voxel was clustered from point cloud or voxel, but the segmentator in ScanNet works on the triangle mesh. If the input is point cloud purely, this segmentator would be useless, right?
